File: Project5_HR/Project5_HR/EmployeeInfoManagement.cpp
#include "../../include/core/EmployeeInfoManagement.h"
#include "../../include/utils/InputValidator.h"
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

EmployeeInfoManagement::EmployeeInfoManagement() 
    : employees(), repository(new FileRepository()), dataFilePath("data/employees.txt") {}

EmployeeInfoManagement::EmployeeInfoManagement(FileRepository* repo)
    : employees(), repository(repo), dataFilePath("data/employees.txt") {}

EmployeeInfoManagement::~EmployeeInfoManagement() {
    if (repository) {
        delete repository;
    }
}

Employee* EmployeeInfoManagement::FindById(int id) {
    for (auto& e : employees) {
        if (e.GetStaffId() == id) return &e;
    }
    return nullptr;
}

void EmployeeInfoManagement::AddStaff(const Employee& e) {
    if (FindById(e.GetStaffId())) {
        cout << "Employee with ID " << e.GetStaffId() << " already exists. Add cancelled.\n";
        return;
    }
    employees.push_back(e);
    cout << "Employee " << e.GetName() << " added successfully.\n";
    SaveToFile();
}

void EmployeeInfoManagement::UpdateStaff(int id) {
    Employee* e = FindById(id);
    if (!e) {
        cout << "Employee with ID " << id << " not found.\n";
        return;
    }

    int choice;
    do {
        cout << "\nWhat do you want to update for employee " << e->GetName() 
             << " (ID " << e->GetStaffId() << ")?\n";
        cout << "1. Name\n";
        cout << "2. Position\n";
        cout << "3. Phone Number\n";
        cout << "4. Hire Date\n";
        cout << "0. Done/Cancel\n";
        
        choice = InputValidator::ReadInt("Enter your choice: ");

        if (choice == 1) {
            string n = InputValidator::ReadString("Enter new name: ");
            e->SetName(n);
            cout << "Name updated.\n";
        }
        else if (choice == 2) {
            string p = InputValidator::ReadString("Enter new position: ");
            e->SetPosition(p);
            cout << "Position updated.\n";
        }
        else if (choice == 3) {
            string num = InputValidator::ReadString("Enter new phone number: ");
            e->SetPhoneNum(num);
            cout << "Phone number updated.\n";
        }
        else if (choice == 4) {
            string d = InputValidator::ReadString("Enter new hire date (YYYY-MM-DD): ");
            e->SetHireDate(d);
            cout << "Hire date updated.\n";
        }
        else if (choice == 0) {
            cout << "Exiting update menu.\n";
        }
        else {
            cout << "Invalid choice. Try again.\n";
        }
    } while (choice != 0);
    
    SaveToFile();
}

Employee* EmployeeInfoManagement::GetStaff(int id) {
    return FindById(id);
}

void EmployeeInfoManagement::RemoveStaff(int id) {
    for (auto it = employees.begin(); it != employees.end(); ++it) {
        if (it->GetStaffId() == id) {
            cout << "Employee " << it->GetName() << " removed.\n";
            employees.erase(it);
            SaveToFile();
            return;
        }
    }
    cout << "Employee with ID " << id << " not found.\n";
}

void EmployeeInfoManagement::DisplayAllEmployees() const {
    cout << "\n==== Employee List ====\n";
    if (employees.empty()) {
        cout << "No employees in the system.\n";
        return;
    }
    for (const auto& e : employees) {
        e.Display();
    }
}

bool EmployeeInfoManagement::EmployeeExists(int id) const {
    for (const auto& e : employees) {
        if (e.GetStaffId() == id) return true;
    }
    return false;
}

void EmployeeInfoManagement::SaveToFile() {
    if (!repository) {
        cerr << "Error: Repository not initialized.\n";
        return;
    }
    
    stringstream ss;
    for (const auto& e : employees) {
        ss << e.Serialize() << "\n";
    }
    
    repository->SaveToFile(ss.str(), dataFilePath);
    cout << "Employee data saved successfully.\n";
}

void EmployeeInfoManagement::LoadFromFile() {
    if (!repository) {
        cerr << "Error: Repository not initialized.\n";
        return;
    }
    
    if (!repository->FileExists(dataFilePath)) {
        cout << "No existing employee data found.\n";
        return;
    }
    
    employees.clear();
    vector<string> lines = repository->ReadLines(dataFilePath);
    
    for (const auto& line : lines) {
        if (!line.empty()) {
            Employee e = Employee::Deserialize(line);
            if (e.GetStaffId() != 0) {
                employees.push_back(e);
            }
        }
    }
    
    cout << "Loaded " << employees.size() << " employees from file.\n";
}

void EmployeeInfoManagement::Management() {
    int choice;
    do {
        cout << "\n==== Employee Info Management System ====\n";
        cout << "1. Add Employee\n";
        cout << "2. Update Employee\n";
        cout << "3. Get Employee\n";
        cout << "4. Remove Employee\n";
        cout << "5. Display All Employees\n";
        cout << "0. Back to Main Menu\n";
        
        choice = InputValidator::ReadInt("Enter your choice: ");

        if (choice == 1) {
            int id = InputValidator::ReadInt("Enter Employee ID (number only): ");
            if (FindById(id)) {
                cout << "Employee with ID " << id << " already exists. Add cancelled.\n";
                continue;
            }
            
            string n = InputValidator::ReadString("Enter name: ");
            string p = InputValidator::ReadString("Enter position: ");
            string num = InputValidator::ReadString("Enter phone number: ");
            string d = InputValidator::ReadString("Enter hire date (YYYY-MM-DD): ");
            
            AddStaff(Employee(id, n, p, num, d));
        }
        else if (choice == 2) {
            int id = InputValidator::ReadInt("Enter Employee ID to update: ");
            UpdateStaff(id);
        }
        else if (choice == 3) {
            int id = InputValidator::ReadInt("Enter Employee ID to get: ");
            Employee* e = GetStaff(id);
            if (e) {
                cout << "\nEmployee Found:\n";
                e->Display();
            }
            else {
                cout << "Employee with ID " << id << " not found.\n";
            }
        }
        else if (choice == 4) {
            int id = InputValidator::ReadInt("Enter Employee ID to remove: ");
            RemoveStaff(id);
        }
        else if (choice == 5) {
            DisplayAllEmployees();
        }
        else if (choice == 0) {
            cout << "Returning to main menu...\n";
        }
        else {
            cout << "Invalid choice! Try again.\n";
        }
    } while (choice != 0);
}

File: Project5_HR_Back_End/header/core/employee_info_management.py
from typing import List, Optional
from employee import Employee
from persistence.file_repository import FileRepository


class EmployeeInfoManagement:
    """
    EmployeeInfoManagement class for managing employee records.

    Attributes:
        employees (list[Employee]): List of employee objects.
        repository (FileRepository): Handles saving/loading data.
        data_file_path (str): Path to employee data file.
    """

    def __init__(self, repo: FileRepository = None):
        """
        Constructor for EmployeeInfoManagement.

        Args:
            repo (FileRepository, optional): Existing repository instance.
        """
        self.employees: List[Employee] = []
        self.repository = repo if repo else FileRepository()
        self.data_file_path = "employees.txt"

    # ----------------------------
    # Internal helper
    # ----------------------------
    def _find_by_id(self, emp_id: int) -> Optional[Employee]:
        """
        Find employee by ID.

        Args:
            emp_id (int): Employee ID.

        Returns:
            Employee | None: Employee object if found.
        """
        for emp in self.employees:
            if emp.staff_id == emp_id:
                return emp
        return None

    # ----------------------------
    # CRUD Operations
    # ----------------------------
    def add_staff(self, e: Employee) -> None:
        """Add a new employee."""
        if self._find_by_id(e.staff_id):
            print(f"Employee with ID {e.staff_id} already exists. Add cancelled.")
            return

        self.employees.append(e)
        print(f"Employee {e.name} added successfully.")
        self.save_to_file()

    def update_staff(self, emp_id: int):
        """Update employee details."""
        e = self._find_by_id(emp_id)
        if not e:
            print(f"Employee with ID {emp_id} not found.")
            return

        while True:
            print(f"\nUpdating employee {e.name} (ID {e.staff_id})")
            print("1. Name")
            print("2. Position")
            print("3. Phone Number")
            print("4. Hire Date")
            print("0. Done/Cancel")

            choice = input("Enter your choice: ")

            if choice == "1":
                e.name = input("Enter new name: ")
                print("Name updated.")
            elif choice == "2":
                e.position = input("Enter new position: ")
                print("Position updated.")
            elif choice == "3":
                e.phone_num = input("Enter new phone number: ")
                print("Phone number updated.")
            elif choice == "4":
                e.hire_date = input("Enter new hire date (YYYY-MM-DD): ")
                print("Hire date updated.")
            elif choice == "0":
                print("Exiting update menu.")
                break
            else:
                print("Invalid choice. Try again.")

        self.save_to_file()

    def get_staff(self, emp_id: int) -> Optional[Employee]:
        """Return an Employee object by ID."""
        return self._find_by_id(emp_id)

    def remove_staff(self, emp_id: int) -> None:
        """Remove employee by ID."""
        for i, e in enumerate(self.employees):
            if e.staff_id == emp_id:
                print(f"Employee {e.name} removed.")
                del self.employees[i]
                self.save_to_file()
                return
        print(f"Employee with ID {emp_id} not found.")

    def display_all_employees(self):
        """Print all employee records."""
        print("\n==== Employee List ====")
        if not self.employees:
            print("No employees in the system.")
            return
        for e in self.employees:
            e.display()

    def employee_exists(self, emp_id: int) -> bool:
        """Check if employee exists."""
        return self._find_by_id(emp_id) is not None

    # ----------------------------
    # File Persistence
    # ----------------------------
    def save_to_file(self):
        """Save all employee data to file."""
        if not self.repository:
            print("Error: Repository not initialized.")
            return

        content = "\n".join(emp.serialize() for emp in self.employees)
        self.repository.save_to_file(content, self.data_file_path)
        print("Employee data saved successfully.")

    def load_from_file(self):
        """Load employees from file."""
        if not self.repository.file_exists(self.data_file_path):
            print("No existing employee data found.")
            return

        self.employees.clear()
        lines = self.repository.read_lines(self.data_file_path)

        for ln in lines:
            if ln.strip():
                emp = Employee.deserialize(ln.strip())
                if emp.staff_id != 0:
                    self.employees.append(emp)

        print(f"Loaded {len(self.employees)} employees from file.")

    # ----------------------------
    # Management Menu
    # ----------------------------
    def management(self):
        """Run interactive console menu."""
        while True:
            print("\n==== Employee Info Management System ====")
            print("1. Add Employee")
            print("2. Update Employee")
            print("3. Get Employee")
            print("4. Remove Employee")
            print("5. Display All Employees")
            print("0. Back to Main Menu")

            choice = input("Enter your choice: ")

            if choice == "1":
                try:
                    emp_id = int(input("Enter Employee ID: "))
                except ValueError:
                    print("Invalid number.")
                    continue

                if self._find_by_id(emp_id):
                    print("Employee already exists.")
                    continue

                name = input("Enter name: ")
                pos = input("Enter position: ")
                phone = input("Enter phone number: ")
                date = input("Enter hire date (YYYY-MM-DD): ")

                self.add_staff(Employee(emp_id, name, pos, phone, date))

            elif choice == "2":
                emp_id = int(input("Enter Employee ID to update: "))
                self.update_staff(emp_id)

            elif choice == "3":
                emp_id = int(input("Enter Employee ID: "))
                emp = self.get_staff(emp_id)
                if emp:
                    emp.display()
                else:
                    print("Employee not found.")

            elif choice == "4":
                emp_id = int(input("Enter Employee ID to remove: "))
                self.remove_staff(emp_id)

            elif choice == "5":
                self.display_all_employees()

            elif choice == "0":
                print("Returning to main menu...")
                break

            else:
                print("Invalid choice, try again.")
